<a href="https://colab.research.google.com/github/Najla-Nuricia/PemMes_244107020091_Najla/blob/main/JS05/JS05_Tugas2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [2]:
from google.colab import files
uploaded = files.upload()

Saving spam.csv to spam.csv


In [3]:
df = pd.read_csv('spam.csv', encoding='latin-1')

df = df[['v1', 'v2']]

df.columns = ['label', 'message']

print(df.head())
print(df.shape)

  label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...
(5572, 2)


In [4]:
X = df['message']
y = df['label']

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.3,
    random_state=42,
    stratify=y
)

print('Jumlah data train:', len(X_train))
print('Jumlah data test:', len(X_test))

Jumlah data train: 3900
Jumlah data test: 1672


In [5]:
bow = CountVectorizer(stop_words='english')

X_train_bow = bow.fit_transform(X_train)
X_test_bow = bow.transform(X_test)

print('Jumlah fitur:', len(bow.get_feature_names_out()))
print('Dimensi data train:', X_train_bow.shape)

Jumlah fitur: 6946
Dimensi data train: (3900, 6946)


In [6]:
mnb_bow = MultinomialNB()

mnb_bow.fit(X_train_bow, y_train)

y_pred_bow = mnb_bow.predict(X_test_bow)

acc_bow = accuracy_score(y_test, y_pred_bow)

print(f'Akurasi CountVectorizer: {acc_bow:.4f}')

Akurasi CountVectorizer: 0.9844


In [7]:
print(classification_report(y_test, y_pred_bow))

              precision    recall  f1-score   support

         ham       0.99      0.99      0.99      1448
        spam       0.95      0.93      0.94       224

    accuracy                           0.98      1672
   macro avg       0.97      0.96      0.97      1672
weighted avg       0.98      0.98      0.98      1672



In [8]:
print(confusion_matrix(y_test, y_pred_bow))

[[1437   11]
 [  15  209]]


berdasarkan hasil tingkat keakuratan yaitu 0.9844

dan confusion matrix untuk data test menghasilkan TN = 1437, FP = 11, FN = 15, dan TP = 209. Artinya, 1437 data negatif diprediksi benar, 11 data negatif salah diprediksi positif, 15 data positif salah diprediksi negatif, dan 209 data positif diprediksi benar.

In [9]:
# TF-IDF dengan stop_words
tfidf = TfidfVectorizer(stop_words='english')

X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

print('Jumlah fitur:', len(tfidf.get_feature_names_out()))
print('Dimensi data train:', X_train_tfidf.shape)

Jumlah fitur: 6946
Dimensi data train: (3900, 6946)


In [10]:
# Inisiasi Multinomial Naive Bayes
mnb_tfidf = MultinomialNB()

# Training
mnb_tfidf.fit(X_train_tfidf, y_train)

# Prediksi
y_pred_tfidf = mnb_tfidf.predict(X_test_tfidf)

# Evaluasi
acc_tfidf = accuracy_score(y_test, y_pred_tfidf)

print(f'Akurasi TF-IDF: {acc_tfidf:.4f}')

Akurasi TF-IDF: 0.9683


In [11]:
print(classification_report(y_test, y_pred_tfidf))

              precision    recall  f1-score   support

         ham       0.96      1.00      0.98      1448
        spam       1.00      0.76      0.87       224

    accuracy                           0.97      1672
   macro avg       0.98      0.88      0.92      1672
weighted avg       0.97      0.97      0.97      1672



In [13]:
confusion_matrix(y_test, y_pred_tfidf)

array([[1448,    0],
       [  53,  171]])

Untuk TF-IDF, confusion matrix menunjukkan TN = 1448, FP = 0, FN = 53, dan TP = 171. Artinya, 1448 data negatif diprediksi benar, tidak ada data negatif yang salah diprediksi positif, 53 data positif salah diprediksi negatif, dan 171 data positif diprediksi benar.

Berdasarkan hasil pengujian pada dataset spam.csv, model Multinomial Naive Bayes dengan fitur CountVectorizer menghasilkan akurasi sebesar 98%, sedangkan penggunaan TF-IDF menghasilkan akurasi sebesar 96%. Dengan demikian,CountVectorizer menghasilkan performa yang lebih tinggi dibandingkan TF-IDF.